### PR TITLE
Substantial improvements on audio quality plus test cases for the Ogg Vorbis decoder

### PR DESCRIPTION
**Pull request by quarterpounder**
Hi,
there are some silly mistakes I'd like to have corrected. The most prominent and now fixed effectively rendered the transcoded Ogg Vorbis audio mono... As I've said, now fixed. Furthermore, there are some test cases for the Ogg Vorbis decoder now, ripped off the existing Flac tests.

### CODE REVIEW

File: src/vorbis_decoder.cc [hunk 1]
```diff
@@ -246,23 +246,24 @@ int VorbisDecoder::process_single_fr(Encoder* encoder, Buffer* buffer) {
...
-                        encode_buffer[channel][i] = decode_buffer.as_int[i];
-                        break;
+                case sizeof(int32_t):
+                    encode_buffer[channel][i] = decode_buffer.as_int[sample];
+                    break;
 
-                    case sizeof(short):
-                        encode_buffer[channel][i] = decode_buffer.as_short[i];
-                        break;
+                case sizeof(short):
```
Comment (khenriks) [inline src/vorbis_decoder.cc]: I don't understand this logic. You set word above to be explicitly equal to sizeof(short).

File: src/vorbis_decoder.cc [hunk 2]
```diff
@@ -246,23 +246,24 @@ int VorbisDecoder::process_single_fr(Encoder* encoder, Buffer* buffer) {
...
+                    encode_buffer[channel][i] = (int32_t)decode_buffer.as_short[sample];
```
Comment (khenriks) [inline src/vorbis_decoder.cc]: In place of the extra `sample` variable, use `i*vi->channels + channel`. Also, I think the cast here is extraneous as it should be done implicitly by the compiler.

File: src/vorbis_decoder.cc [hunk 3]
```diff
@@ -273,22 +274,17 @@ int VorbisDecoder::process_single_fr(Encoder* encoder, Buffer* buffer) {
...
         for (int channel = 0; channel < vi->channels; ++channel) {
-            if (encode_buffer[channel] != NULL) {
-                delete[] encode_buffer[channel];
-                encode_buffer[channel] = NULL;
-            }
+            delete[] encode_buffer[channel];
+            encode_buffer[channel] = NULL;
         }
-
+
```
Comment (khenriks) [inline src/vorbis_decoder.cc]: No lines ending in whitespace

File: test/test_tags_vorbis [hunk 4]
```diff
@@ -14,5 +14,5 @@ TLEN=6000
 TPE1=Barack Obama
 TPOS=0/0
 TRCK=1/1
-TSSE=MP3FS
+TSSE=MP3FS
```
Comment (khenriks) [inline test/test_tags_vorbis]: Trailing whitespace

File: src/vorbis_decoder.cc (hunk 1 above)
Comment (quarterpounder) [inline src/vorbis_decoder.cc]: As there's no need anymore to keep the decoder reconfigurable, the switch statement can be removed now.

File: src/vorbis_decoder.cc (hunk 2 above)
Comment (quarterpounder) [inline src/vorbis_decoder.cc]: D'accord.

File: test/test_tags_vorbis (hunk 4 above)
Comment (quarterpounder) [inline test/test_tags_vorbis]: Can you please run 'make check' with and without this trailing whitespace?

File: src/vorbis_decoder.cc (hunk 3 above)
Comment (quarterpounder) [inline src/vorbis_decoder.cc]: D'accord.

**Comment by quarterpounder** [comment]
Hello khenriks,

this pull request stabilizes the support for Ogg Vorbis, adds common automated tests and fixes the automated build test (f___ Ubuntu wouldn't ship with libb64). So please consider it a merge candidate.

Yours kindly,
quarterpounder

**Comment by khenriks** [comment]
Hi quarterpounder,

After a bit of a hiatus, I've been working on mp3fs more, and I had actually completely forgotten about this pull request. I'd been working on investigating and fixing essentially the same thing.

For the testing, I wanted to take a slightly different approach, but since you've done the work to fix the Vorbis decoder, I'd like to accept that part of the request.

Can you separate out just the changes to vorbis_decoder.cc in this request? If you can also squash the changes into a single commit, that would be great.

Regards,
khenriks

**Comment by quarterpounder** [comment]
Hi khenriks,

a quick git diff gave me this:
[changes-from-#40.zip](https://github.com/khenriks/mp3fs/files/979112/changes-from-.40.zip)

Maybe, this is helpful.

Cheers,
quarterpounder

**Comment by khenriks** [comment]
I can grab the diffs myself, I just thought I'd make it a proper pull request so it had your name attached and whatnot. If it's not important to you though, I'll just apply the diffs for that file directly then.

**Comment by khenriks** [comment]
These improvements have already been applied separately at master, so closing.

**Pull request closed without merging**